crates/cxx-qt-gen/src/preprocessor/self_inlining.rs

```diff
@@ -0,0 +1,191 @@
+// SPDX-FileContributor: Ben Ford <ben.ford@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::parser::cxxqtdata::ParsedCxxQtData;
+use crate::parser::inherit::ParsedInheritedMethod;
+use crate::parser::method::{MethodFields, ParsedMethod};
+use crate::parser::qobject::ParsedQObject;
+use crate::parser::signals::ParsedSignal;
+use crate::Parser;
+use proc_macro2::Ident;
+use quote::format_ident;
+use std::ops::DerefMut;
+use syn::spanned::Spanned;
+use syn::{Error, Result};
+
+/// Inline any `Self` types in the methods signatures with the Ident of a qobject passed in
+///
+/// If there are unresolved methods in the list, but inline is false, it will error,
+/// as the self inlining is only available if there is exactly one `QObject` in the block,
+/// and this indicates that no inlining can be done, but some `Self` types were present.
+pub fn try_inline_self_invokables(
+    inline: bool,
+    type_to_inline: &Option<Ident>,
+    invokables: &mut [impl DerefMut<Target = MethodFields>],
+) -> syn::Result<()> {
+    for method in invokables.iter_mut() {
+        if method.self_unresolved {
+            if inline {
+                if let Some(inline_type) = type_to_inline.clone() {
+                    method.qobject_ident = inline_type;
+                } else {
+                    return Err(Error::new(
+                        method.method.span(),
+                        "Expected a type to inline, no `qobject` typename was passed!",
+                    ));
+                }
+            } else {
+                return Err(Error::new(
+                    method.method.span(),
+                    "`Self` type can only be inferred if the extern block contains only one `qobject`.",
+                ));
+            }
+        }
+    }
+    Ok(())
+}
+
+/// A collection of items found in an extern block when parsing
+type BlockComponents<'a> = (
+    &'a mut Vec<ParsedQObject>,
+    &'a mut Vec<ParsedMethod>,
+    &'a mut Vec<ParsedInheritedMethod>,
+    &'a mut Vec<ParsedSignal>,
+);
+
+// Separates the parsed data by block and returns tuples of the components
+fn separate_blocks(data: &mut ParsedCxxQtData) -> Vec<BlockComponents> {
+    data.qobjects
+        .iter_mut()
+        .zip(data.methods.iter_mut())
+        .zip(data.inherited_methods.iter_mut())
+        .zip(data.signals.iter_mut())
+        .map(|(((qobject, method), inherited_method), signal)| {
+            (qobject, method, inherited_method, signal)
+        })
+        .collect()
+}
+
+/// For a given parser, attempt to inline the `Self` type used in any of the blocks with that blocks unique QObject
+pub fn qualify_self_types(parser: &mut Parser) -> Result<()> {
+    for (qobjects, methods, inherited, signals) in separate_blocks(&mut parser.cxx_qt_data) {
+        let inline_self = qobjects.len() == 1;
+        let inline_ident = qobjects
+            .last()
+            .map(|obj| format_ident!("{}", obj.declaration.ident_left));
```

Comment:
The `inline_self` feels like it's carrying redundant information.
The `Option` should either be `Some`, in which case we can inline, or `None` if we cannot inline.

How about getting rid of `inline_self` and instead doing:
```rust
let mut qobjects = qobjects.iter();
let mut inline_ident = qobjects.next().map(|obj| obj.declaration.ident_left.clone());
if qobjects.next().is_some() {
    // More than one ident found
    inline_ident = None;
}
```